### PR TITLE
Hypervisor extension integration

### PR DESCRIPTION
**Pull request by andresag01**
This is a first shot at integrating RISC-V's hypervisor (H) extension with CHERI. At a high-level, my perspective is that CHERI is generally orthogonal from hypervisor, so most of the changes here are limited to extending the hypervisor-related address CSRs to capability size, adding enable bits and taking care of XLEN configurations.

**NOTE:** This PR is currently a draft to get the discussion started and collect early feedback about the approach taken here. There are a few bits missing before this can be merged like updating tables, cross references, etc.

@tariqkurd-repo @jrtc27 @arichardson : What do you think?

### CODE REVIEW

File: src/hypervisor-integration.adoc [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+supports {cheri_legacy_ext_name} as shown in xref:henvcfgreg[xrefstyle=short].
+
+.Hypervisor environment configuration register (*henvcfg*)
+[#henvcfgreg]
+include::img/henvcfgreg.edn[]
+
+The CHERI Register Enable (<<CRE>>) bit controls whether explicit access to
+CHERI register is permitted when V=1. When <<CRE>>=1, CHERI registers
+can be read and written by VS and VU modes. When <<CRE>>=0, CHERI registers are
+disabled in VS and VU modes.
```
Comment (PRugg-Cap) [inline src/hypervisor-integration.adoc:40]: Presumably the intent is that the less privileged modes all read their CRE bits as 0 as well? I.e. henvcfg.cre gets added to the cross-CSR wall field?

**Review by PRugg-Cap — COMMENTED** [review]
Looks like the right approach. I wonder if we need a HTID(c) (@francislaus?).

File: src/hypervisor-integration.adoc (hunk 1 above)
Comment (andresag01) [reply src/hypervisor-integration.adoc:40]: This was recently changed and the CRE bits are no longer a WARL

**Comment by andresag01** [comment]
> Looks like the right approach. I wonder if we need a HTID(c) (@francislaus?).

Ah good point. I will wait for @francislaus to comment.

**Comment by andresag01** [comment]
@tariqkurd-repo @arichardson : I reworked the PR. It is now up-to-date with master and conflict-free. I also added missing CSRs and hypervisor instructions that need CHERI support.

Now we are only missing:
* Allocating encodings for HLV.C and HSV.C
* Updating the auto-generated tables from the spreadsheet
    * I would prefer to do this at the end. Otherwise the changes will get pulled in every other PR when the spreadsheet is updated!
    * Reviews!

NOTE: I found a few other things I fixed along the way elsewhere in the spec. I will eventually split those commits out of this PR when I get some time.

**Comment by andresag01** [comment]
Allocated encodings for HLV.C and HSV.C

File: src/insns/sh123add_32bit.adoc [hunk 2]
```diff
@@ -28,7 +28,7 @@ Encoding::
 {reg:[
     { bits:  7, name: 0x33, attr: ['OP'] },
     { bits:  5, name: 'cd/rd' },
-    { bits:  3, name: 0x2, attr: ['SH1ADD=010', 'SH2ADD=100', 'SH3ADD=110'] },
+    { bits:  3, name: 'func3', attr: ['SH1ADD=010', 'SH2ADD=100', 'SH3ADD=110'] },
```
Comment (arichardson) [inline src/insns/sh123add_32bit.adoc:31]: Unrelated change, probably best to submit the separately.

File: src/insns/sh123adduw_32bit.adoc [hunk 3]
```diff
@@ -28,7 +28,7 @@ Encoding::
 {reg:[
     { bits:  7, name: 0x3b, attr: ['OP'] },
     { bits:  5, name: 'cd/rd' },
-    { bits:  3, name: 0x2, attr: ['rv64: SH1ADD.UW=010', 'rv64: SH2ADD.UW=100', 'rv64: SH3ADD.UW=110'] },
+    { bits:  3, name: 'func3', attr: ['rv64: SH1ADD.UW=010', 'rv64: SH2ADD.UW=100', 'rv64: SH3ADD.UW=110'] },
```
Comment (arichardson) [inline src/insns/sh123adduw_32bit.adoc:31]: Also unrelated.

File: src/riscv-hybrid-integration.adoc [hunk 4]
```diff
@@ -199,9 +199,9 @@ Load and store instructions change behavior depending on the CHERI execution
 mode although the instruction's encoding remains unchanged.
 
 Loads and stores behave as described in
-xref:section_existing_riscv_insns[xrefstyle=short] when in pass:attributes,quotes[{cheri_cap_mode_name}]
+xref:section_existing_riscv_insns[xrefstyle=short] when in pass:attributes,quotes[{cheri_cap_mode_name}].
 In pass:attributes,quotes[{cheri_int_mode_name}], the instructions behave as described in
-the RISC-V base ISA (i.e. without the 'C' prefix) and rely on *x* operands
+the RISC-V base ISA and rely on *x* operands
```
Comment (arichardson) [inline src/riscv-hybrid-integration.adoc:204]: Same here, would be cleaner to commit separately.

File: src/riscv-integration.adoc [hunk 5]
```diff
@@ -424,11 +424,11 @@ Only 1 and 2 are supported values for SXL and UXL and the fields must be
 read-only in implementations supporting {cheri_base_ext_name}. The effective
 XLEN in S-mode and U-mode are termed SXLEN and UXLEN, respectively.
 
-The MBE, SBE, and UBE fields determine whether explicit loads and stores
-performed from M-mode, S-mode, or U-mode, respectively, are little endian
-(xBE = 0) or big endian (xBE = 1).  MBE must be read only.  SBE and UBE must be
-read only and equal to MBE, if S-mode or U-mode, respectively, is implemented,
-or read only zero otherwise.
+The MBE, SBE, and UBE fields determines the endianness of memory accesses other
```
Comment (arichardson) [inline src/riscv-integration.adoc:427]:
```suggestion
The MBE, SBE, and UBE fields determine the endianness of memory accesses other
```

**Review by arichardson — APPROVED** [review]
I think this looks great and we should go ahead and merge it.

There are some unrelated changes but as long as we don't squash this PR into one commit and retain the history that is fine.

**Comment by andresag01** [comment]
@arichardson : Removed all commits that were unrelated to the hypervisor changes.

**Comment by andresag01** [comment]
@arichardson : Thanks for reviewing. I added the hypervisor CSRs and instructions into the GDrive spreadsheet so the generated tables are now up to data. Also, I added a vstid[c] copy (cc. @PeterRugg). This is ready for final review and merge, so I removed the DRAFT label.

**Review by arichardson — APPROVED** [review]

File: src/hypervisor-integration.adoc [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+[#vsepcc,reftext="vsepcc"]
+=== Virtual Supervisor Exception Program Counter Capability (vsepcc)
+
+The <<vsepcc>> register is a renamed extension of <<vsepc>> that is able to
+hold a capability. Its reset value is the <<infinite-cap>> capability.
+
+As shown in xref:CSR_exevectors[xrefstyle=short], <<vsepcc>> is an executable
+vector, so it need not be able to hold all possible invalid addresses.
+Additionally, the capability in <<vsepcc>> is unsealed when it is installed in
+<<pcc>> on execute of an <<SRET>> instruction. The handling of <<vsepcc>> is
```
Comment (tariqkurd-repo) [inline src/hypervisor-integration.adoc:140]:
```suggestion
<<pcc>> on execution of an <<SRET>> instruction. The handling of <<vsepcc>> is
```

File: src/insns/wavedrom/hypv-virt-load.adoc [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[wavedrom, ,svg]
+....
+{reg: [
+  {bits: 7,  name: 0x73,      attr: ['7'], type: 8},
+  {bits: 5,  name: 'rd',      attr: ['5'], type: 2},
+  {bits: 3,  name: 0x4,       attr: ['3'], type: 8},
+  {bits: 5,  name: 'rs1/cs1', attr: ['5', 'src1'], type: 4},
+  {bits: 5,  name: 'type',    attr: ['5', 'HLV.B=00000', 'HLV.BU=00001', 'HLV.H=00000', 'HLV.HU=00001', 'HLV.W=00000', 'HLV.WU=00001', 'HLV.D=00000'], type: 3},
+  {bits: 7,  name: 'funct7',  attr: ['7', 'width', 'HLV.B=0110000', 'HLV.BU=0110000', 'HLV.H=0110010', 'HLV.HU=0110010', 'HLV.W=0110100', 'HLV.WU=0110100', 'HLV.D=0110110'], type: 3},
```
Comment (tariqkurd-repo) [inline src/insns/wavedrom/hypv-virt-load.adoc:9]:
```suggestion
  {bits: 7,  name: 'funct7',  attr: ['7', 'HLV.B=0110000', 'HLV.BU=0110000', 'HLV.H=0110010', 'HLV.HU=0110010', 'HLV.W=0110100', 'HLV.WU=0110100', 'HLV.D=0110110'], type: 3},
```

File: src/insns/wavedrom/hypv-virt-load-cap.adoc [hunk 8]
```diff
@@ -0,0 +1,11 @@
+[wavedrom, ,svg]
+....
+{reg: [
+  {bits: 7,  name: 0x73,      attr: ['7'], type: 8},
+  {bits: 5,  name: 'cd',      attr: ['5'], type: 2},
+  {bits: 3,  name: 0x4,       attr: ['3'], type: 8},
+  {bits: 5,  name: 'rs1/cs1', attr: ['5', 'src1'], type: 4},
+  {bits: 5,  name: 'type',    attr: ['5', 'HLV.C=00000'], type: 3},
+  {bits: 7,  name: 'funct7',  attr: ['7', 'width', 'HLV.C=0111000'], type: 3},
```
Comment (tariqkurd-repo) [inline src/insns/wavedrom/hypv-virt-load-cap.adoc:9]:
```suggestion
  {bits: 7,  name: 'funct7',  attr: ['7', 'HLV.C=0111000'], type: 3},
```

File: src/insns/wavedrom/hypv-virt-loadx.adoc [hunk 9]
```diff
@@ -0,0 +1,11 @@
+[wavedrom, ,svg]
+....
+{reg: [
+  {bits: 7,  name: 0x73,      attr: ['7'], type: 8},
+  {bits: 5,  name: 'rd',      attr: ['5', 'dest'], type: 2},
+  {bits: 3,  name: 0x4,       attr: ['3'], type: 8},
+  {bits: 5,  name: 'rs1/cs1', attr: ['5', 'src1'], type: 4},
+  {bits: 5,  name: 0x3,       attr: ['5'], type: 3},
+  {bits: 7,  name: 'funct7',  attr: ['7', 'width', 'HLVX.HU=0110010', 'HLVX.WU=0110100'], type: 3},
```
Comment (tariqkurd-repo) [inline src/insns/wavedrom/hypv-virt-loadx.adoc:9]:
```suggestion
  {bits: 7,  name: 'funct7',  attr: ['7', 'HLVX.HU=0110010', 'HLVX.WU=0110100'], type: 3},
```

File: src/insns/wavedrom/hypv-virt-store-cap.adoc [hunk 10]
```diff
@@ -0,0 +1,11 @@
+[wavedrom, ,svg]
+....
+{reg: [
+  {bits: 7,  name: 0x73,      attr: ['7'], type: 8},
+  {bits: 5,  name: 0x0,       attr: ['5'], type: 2},
+  {bits: 3,  name: 0x4,       attr: ['3'], type: 8},
+  {bits: 5,  name: 'rs1/cs1', attr: ['5', 'src1'], type: 4},
+  {bits: 5,  name: 'cs2',     attr: ['5', 'src2'], type: 3},
+  {bits: 7,  name: 'funct7',  attr: ['7', 'width', 'HSV.C=0111001'], type: 3},
```
Comment (tariqkurd-repo) [inline src/insns/wavedrom/hypv-virt-store-cap.adoc:9]:
```suggestion
  {bits: 7,  name: 'funct7',  attr: ['7', 'HSV.C=0111001'], type: 3},
```

File: src/insns/wavedrom/hypv-virt-store.adoc [hunk 11]
```diff
@@ -0,0 +1,11 @@
+[wavedrom, ,svg]
+....
+{reg: [
+  {bits: 7,  name: 0x73,      attr: ['7'], type: 8},
+  {bits: 5,  name: 0x0,       attr: ['5'], type: 2},
+  {bits: 3,  name: 0x4,       attr: ['3'], type: 8},
+  {bits: 5,  name: 'rs1/cs1', attr: ['5', 'src1'], type: 4},
+  {bits: 5,  name: 'rs2',     attr: ['5', 'src2'], type: 3},
+  {bits: 7,  name: 'funct7',  attr: ['7', 'width', 'HSV.B=0110001', 'HSV.H=0110011', 'HSV.W=0110101', 'HSV.D=0110111'], type: 3},
```
Comment (tariqkurd-repo) [inline src/insns/wavedrom/hypv-virt-store.adoc:9]:
```suggestion
  {bits: 7,  name: 'funct7',  attr: ['7', 'HSV.B=0110001', 'HSV.H=0110011', 'HSV.W=0110101', 'HSV.D=0110111'], type: 3},
```

File: src/hypervisor-integration.adoc [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+{REQUIRE_CRE_CSR}
+
+.Virtual supervisor trap default capability register
+include::img/vstdcreg.edn[]
+
+=== Existing Hypervisor Load and Store Instructions
+
+The hypervisor extension defines several integer load and store instruction
+(HLV.*x*, HSV.*x* and HLVX.*x*) that transfer the amount of integer data
```
Comment (tariqkurd-repo) [inline src/hypervisor-integration.adoc:170]: it would be better to have xrefs to the instructions here, although there are several, but to make it clear that the instruction pages do exist

Comment (andresag01) [reply src/hypervisor-integration.adoc]: Fixed
